lib/winapi/profiling.c

```diff
@@ -3,22 +3,59 @@
 // SPDX-FileCopyrightText: 2019 Stefan Schmidt
 
 #include <profileapi.h>
+#ifdef USE_RDTSC_FOR_FREQ
+#include <synchapi.h>
+#endif
 #include <assert.h>
 #include <stddef.h>
 #include <xboxkrnl/xboxkrnl.h>
 
+#ifdef USE_RDTSC_FOR_FREQ
+static LARGE_INTEGER frequency = {{0, 0}};
+static void __attribute__((constructor)) PrimeQueryPerformanceFrequency ()
+{
+    ULARGE_INTEGER f_rdtsc = {{0, 0}}, s_rdtsc = {{0, 0}};
+    ULONG f_ticks = 0, s_ticks = 0;
+
+    KeEnterCriticalRegion();
+
+    // The values generated after launching aren't accurate, give it time to increment...
+    Sleep(700);
```

Comment:
> The values generated after launching aren't accurate

This can't be true for rdtsc. I'm not sure about the clock source for the KeTickCount, but I assume it's some other low level timer.

Also, why would the ticks be incorrect during early boot?! Especially in an XBE I'd expect a soft-reset.

By the point the XBE loads, a bunch of kernel code has ran already. The `rdtsc` is a CPU hw feature and directly tied to the clock in some older x86 CPUs (including Pentium 3 I assume?), I've never heard of it being unstable (at least not for 700ms) - if it was, then your CPU would likely hang.

If your reference clock source (`KeTickCount`) is unstable, then find one which is. However, I assume if your CPU clock is off, then you can't really trust the other clocks either. I also don't think `KeTickCount` has such a bad startup time either.

All of this hackery, to me, looks like it's only meant to support overclocking the CPU (so you attempt to measure a constant factor of your CPU clock to some reference clock which runs at the stock clock).